racket/collects/racket/private/dict.rkt

```diff
@@ -159,23 +160,23 @@
 (define (assoc-count d)
   (unless (assoc? d)
     (raise-argument-error 'dict-count "dict?" d))
-  (length d))
+  (length (assoc-keys d)))
 
 (struct assoc-iter (head pos))
 
 (define (assoc-iterate-first d)
   (unless (assoc? d)
     (raise-argument-error 'dict-iterate-first "dict?" d))
-  (if (null? d) #f (assoc-iter d d)))
+  (if (null? d) #f (assoc-iter d (assoc->list d))))
 
 (define (assoc-iterate-next d i)
   (cond
     [(and (assoc-iter? i)
-          (eq? d (assoc-iter-head i)))
+          (equal? d (assoc-iter-head i)))
```

Comment:
I don't think I understand the reason for changing `eq?` to `equal?` here. I think `(dict-iterate-next d i)` is only allowed with an `i` from a previous call to `(dict-iterate-first d)` for an `eq?` `d`, so I'm not sure why the potentially-very-expensive structural equality check would be needed.

Reply:
This was a misunderstanding on my side of how the code works. I've reverted the change. Thanks.